### ISSUE TITLE
Calculate future positions of masses using orbital elements

### ISSUE THREAD

**Issue by TheHappyKoala** [TheHappyKoala/Harmony-of-the-Spheres#115]

**Comment by TheHappyKoala** [comment]
Happy almost Monday 😄 ! 

I remember you'd written some code for the Lambert solver that would stop it from calculating a trajectory that would pass through the planet the spacecraft is departing from, since, generally speaking, such trajectories do not make for healthy spacecraft 😅. 

I'm asking because chances are I've consigned that code to the dust bin of history with some of the changes that I made to the solver last summer. Alternatively, that check is only done with respect to the planet whose sphere of influence the spacecraft is in, the issue being that when travelling from Earth to, say, Mars, the trajectory is actually calculated with respect to the Sun (yields much more accurate results, but when calculating a trajectory to the Moon, you do so with respect to Earth, still). 

Not sure I'm making much sense here, but if there's a chance you remember anything related to the above, lemme know 😄 . Aint a big issue but I've probably lost 120 crews today alone 😭

**Comment by HugoGranstrom** [comment]
I'm afraid I never wrote any code that did that sadly 😅 I may have said that we would eventually need it (which we do obviously 😂) but no code was ever written. My reasoning was along the lines of that the nearest distance to the starting body (perihelion) of the orbit must be greater than the starting body's radius. But I never came any longer with that idea 😅

**Comment by HugoGranstrom** [comment]
So if we want to calculate the velocities with respect to the Sun but we start from the Earth, then we could get into this problem. My main concern is how we calculate the shortest distance to the starting body if the orbit is hyperbolic with respect to it

**Comment by HugoGranstrom** [comment]
I'm not sure how we would solve it if we does find a way either 😅 I think we have one variable to play with (how many revolutions around the sun before we arrive)

**Comment by TheHappyKoala** [comment]
Hahahaha, ah, got ya! 

The risk of this happening is only high enough to be bothered about when you are in a really tight orbit around the planet, so what we could implement is a simple brute force check, and if that check determines that there will be collision, you simply display a validation message to the user telling them that they have to try a trajectory with a different tof or depart from a different point in the orbit.

**Comment by HugoGranstrom** [comment]
Found a formula we perhaps could use here: https://en.m.wikipedia.org/wiki/Hyperbolic_trajectory#Impact_parameter_and_the_distance_of_closest_approach

**Comment by HugoGranstrom** [comment]
Yeah that seems like a sane way to handle it 😄👍😊

**Comment by TheHappyKoala** [comment]
<img width="657" alt="Screen Shot 2020-02-02 at 18 04 18" src="https://user-images.githubusercontent.com/23419647/73611847-88509b00-45e6-11ea-97c6-bc0751fefeaa.png">

The formula is probably the way to go, but I'll deal with that once the orbital insertion controls are done and dusted... Don't want to get distracted - happens very easily to me, I have the attention span of a 🐶!

On the subject of getting distracted... Look, our first procedurally generated planet! Nothing fancy, but soon enough their appearance will be determined by their temperature, mass and so on... Cool times ahead 😄 

Have a mighty good start to February... Almost spring now 🌞 !

**Comment by TheHappyKoala** [comment]
<img width="595" alt="Screen Shot 2020-02-02 at 18 15 01" src="https://user-images.githubusercontent.com/23419647/73611984-006b9080-45e8-11ea-9c09-03a3da34a6d9.png">

The white are ocean basis / lava plains... Clearly I'm yet to figure out how to work with colors 🤣

**Comment by TheHappyKoala** [comment]
And we need procedural gas giants, neptunes and what not!

**Comment by TheHappyKoala** [comment]
<img width="742" alt="Screen Shot 2020-02-03 at 21 22 17" src="https://user-images.githubusercontent.com/23419647/73687967-611fc980-46cb-11ea-8ee9-0ddb2c0eda36.png">

Man, this is so cool I'm shitting ma pants!

**Comment by HugoGranstrom** [comment]
😮😮😮😮😮😮😮😮🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩 THAT'S BLOODY BRILLIANT!!! 🌎🌍🌏🚀🚀🚀🚀🚀🚀 It looks so good! Is there algorithms for doing this or what's going on?

**Comment by TheHappyKoala** [comment]
It's gonna take the exoplanet scenarios to a whole other level :D But yes, there are! I'm using something called perlin noise subject to the constraints I've already mentioned; temperature, mass and so on. 

Gonna make a big commit this weekend... If you like algorithms I think you'll enjoy reading through the code 😄

**Comment by HugoGranstrom** [comment]
Yes it really will take it to a new level 🚀😉👍 

Gonna check it out, and YES you can bet I'm going to enjoy reading it 😁 

Haven't really started with the Orbital Elements yet 😅 I've been prioritizing  working through the exercises in my programming course, but now the next one is due at the end of February so I'll do my best not to procastinate i anymore 😅😂

**Comment by TheHappyKoala** [comment]
No rush! I work at a very slow pace, too. And I just realised I'm shit posting in your issue... Sorry 😇

**Comment by HugoGranstrom** [comment]
Nah, this is how we roll 😎 we post our shit wherever we like 😉

**Comment by HugoGranstrom** [comment]
Finally the puzzle pieces in my head has fallen into place! 🤣 I have found formulas for converting to and from orbital elements [here](https://downloads.rene-schwarz.com/download/M002-Cartesian_State_Vectors_to_Keplerian_Orbit_Elements.pdf) and [here](https://downloads.rene-schwarz.com/download/M001-Keplerian_Orbit_Elements_to_Cartesian_State_Vectors.pdf). And on [Wikipedia](https://en.wikipedia.org/wiki/Orbital_elements#Orbit_prediction) I found how to propagate them forward in time. So now it's just a matter of implementing them. 

I'm thinking of creating a TS type called OrbitalElements so we don't have to specify all of them as inputs to our functions. What do you think about that? 😄

**Comment by HugoGranstrom** [comment]
Is there anything from your branches that could cause any headache-inducing merge conflicts or am I free to do whatever I want in `lambert.ts`? 😅

**Comment by TheHappyKoala** [comment]
Go for it, and yay 😄 !!! That will make the space flight stuff so much more interesting since you won't have to wait for the trajectory cruncher to generate the future position of a mass. Maybe we could add a date control, too, so that people could view the solar system at different points in time???

**Comment by TheHappyKoala** [comment]
And I really need to merge the gatsby branch soon... It's turning into a monster, haha

**Comment by HugoGranstrom** [comment]
Haha yeah it takes quite some time right now 😅🤣 
The date control should work, but I don't know how accurate it would be for longer timescales 😅 My Earth is already off by 0.02 AU after one year. Don't know how the numbers look for our integrators but they might be off by the same 🤣 It would be a neat feature nonetheless 😉 It wouldn't be to hard to do once we can do it for one mass

**Comment by HugoGranstrom** [comment]
We have our monster of our own we could use to tame Gatsby: RKN12 🤣 🤣

**Comment by TheHappyKoala** [comment]
I should have left it at implementing Gatsby but instead I doubled down on the gitigans by baking other features like new textures, procedural stars etc into it... Will I ever learn 👽 ? 

Yea, the farther into the future, or past for that matter, you try to ascertain the position of a planet, the less accurate it will be, especially for comets that are often perturbed by Jupiter. Would be kind of cool to create one scenario where you have a set of clones running on orbital elements and compare their positions to a set of planets whose positions are determined by an integrator and get a sense of how important perturbations are to our solar system. 

Yes, we should wire reality up to RKN12 and make things happen  🤣 .

**Comment by HugoGranstrom** [comment]
Gitigans have a certain charm to them 😂 it's hard to resist sometimes. You have gotten loads of work done nonetheless 😉 

Oh yeah, that would be super cool 🚀🚀😄😄😄😎 I think we would pretty quickly see the effects for certain bodies (mercury) and some would probably not change much (Neptune)

**Comment by HugoGranstrom** [comment]
Do you prefer to have the Orbital Elements in degrees or radians? For doing all the future telling it's more convenient to have them in radians all along. Do you have any opinions on the matter? 😄

**Comment by TheHappyKoala** [comment]
Nah, doesn't really matter. For whatever reason I have a natural preference for degrees (guess they feel more familliar!), but at the end of the day converting between the two is a piece of cake. Fact of the matter is that switching to radians might make more sense seeing as THREE.js uses radians and not degrees 🚀 !

**Comment by TheHappyKoala** [comment]
First commit for the procedural planets is on its way... Just the terrestial ones, but they are trickier to generate than the gas giants, so there's still a yay factor to this commit, I'd say 🤣

**Comment by HugoGranstrom** [comment]
Radians it is then! 😉 Felt better getting your opinion on this so I didn't suddenly change it to radians and all your code assuming degrees stopped working 🤣 

Awesome! 🚀 Looking forward to populating a whole galaxy with procedural planets 🤣

**Comment by HugoGranstrom** [comment]
I have btw managed to get the stateToKepler, keplerToState and propagateOrbitalElements to work so now I just have to make functions for traversing a SOITree and doing all the propagation for all bodies.

I'm thinking of making two different functions:
- One we will use in the trajectory cruncher that only cares about the future position of one body. It will propagate the body and all it's parents forward to get the final position. Ex Titan: first spin Titan around Saturn and then Saturn around the Sun.
- The other function calculates the future positions of ALL bodies in the system so it does the job the integrators would have otherwise done. 

Not sure how to handle the particles though... 🤔

**Comment by TheHappyKoala** [comment]
Ohhhh... Let me get back to you tomorrow with some feedback regarding the points you brought up...  Right now I'm obsessively trying to wrap up the procedural planets... So close but there's this little issue pertaining to the interpolation of colors that is driving me nuuuuuuts 🤣

**Comment by HugoGranstrom** [comment]
No hurry 😉👍 

Oh that is so annoying when there's just one small thing left and it's not cooperating 😔😦🦕

**Comment by TheHappyKoala** [comment]
<img width="575" alt="Screen Shot 2020-02-14 at 19 07 23" src="https://user-images.githubusercontent.com/23419647/74556053-57b11f80-4f5d-11ea-8830-b1f4716c0c66.png">

LOL. Search terms that brought people to our website from google... One of those guys must have been VERY disappointed when he saw what our site was actually about 🤣 . I mean, wtf! Should I laugh or loose faith in humanity?

**Comment by HugoGranstrom** [comment]
Hahahahaha, what the #$_* 🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣 An idea for the future perhaps?

**Comment by TheHappyKoala** [comment]
X-Rated HOS... 🤣 No, somebody will have to fork HOS and create it themselves

**Comment by TheHappyKoala** [comment]
I would love it if HOS was the basis of the world's first, ehrm, 3D space sex simulator

**Comment by HugoGranstrom** [comment]
🤣🤣 I would be hugely impressed if someone managed that feat

**Comment by TheHappyKoala** [comment]
Totally. How you doing? Any extravagant plans for the weekend?

**Comment by HugoGranstrom** [comment]
I'm doing fine, school is really chill right now and life is rolling on 😄 the weather does seem too nice in the weekend so hopefully I will get some coding done. How are you and what's your plan? 😊

**Comment by TheHappyKoala** [comment]
I'm doing pretty good. Work's still very challening and a lot of fun, and on my way home today I bought two beers at bolaget, so tonight I'm gonna drink beer and work on the procedural planets 🤣 I'm such a bad boy.

**Comment by HugoGranstrom** [comment]
Sounds great!😄👍 Haha maybe a more "open" mind can be helpful there 😅🤣

**Comment by TheHappyKoala** [comment]
![Screen Shot 2020-02-15 at 21 03 10](https://user-images.githubusercontent.com/23419647/74594425-a67eb800-5036-11ea-92a1-2f81074ac804.png)

**Comment by TheHappyKoala** [comment]
Creating world's is so much fun!!! This is a cool pic because it is the first time I have generated a planet based on its temperature, mass and radius

**Comment by HugoGranstrom** [comment]
Wow! 😮🤩🤩🤩 That looks so realistic but still a bit random 😉 what's thr difference between green and brown?

**Comment by TheHappyKoala** [comment]
Brown is supposed to represent mountains...  I need a more realistic model for how vegetation changes with elevation. Habitable planets are tricky... World's like Mars and Mercury are more gratifying 🤣.  Also, I need to take the longitude of a pixel into account, so that poles are covered by ice caps,  whereas somewhat above and below the equator you get desserts.

**Comment by HugoGranstrom** [comment]
Ahhh I get it 😄👍 Sounds like a lot of work 😅

**Comment by TheHappyKoala** [comment]
Hms, no idea why I hadn't thought of this before, but by passing a uniform to the shader and calculating the temperature (black body, taking greenhouse gases into account is beyond ma paygrade) of Earth based on its distance from the Sun, I can change the look of Earth... So take it out into the asteroid belt and it freezes over, and inwards towards Mercury and it becomes a dessert world. That'll have to wait for later, though... Need to make small and manageable commits 😅

**Comment by TheHappyKoala** [comment]
OMG, OMG, OMG!!!

You know how to make sure that some property on an object is not undefined or null, you have to do a check like this:  

```
data !== null && data.person && data.person.name && doSomething(data.person.name);
```

Well... With the newest version of JavaScript, you can do the same check like so:

```
data?.person?.name? doSomething(data.person.name);
```

🎈 💯 🏅 🎉

**Comment by HugoGranstrom** [comment]
Wow, that's such a brilliant idea in its simplicity 😉😊😎 Seeing how Earth changes depending on the distance to the Sun

Didn't actually know you had to go through all that trouble tbh 🤣😂🤣😂 but the new way is much neater and shorter, guess it saves you a lot of time?😊

**Comment by TheHappyKoala** [comment]
Doesn't save you that much time, it's more the experience of reading code that improves... It's the difference between caligraphy and writing like this "hOw u dUiN?!!" with bright red marker 🤣

**Comment by TheHappyKoala** [comment]
<img width="1002" alt="Screen Shot 2020-02-17 at 21 07 46" src="https://user-images.githubusercontent.com/23419647/74683551-a105a700-51c9-11ea-8484-cf473ead82b4.png">

Maaannn.... So close now, even figured out how to make procedural gas giants!!! Gonna tweak the model that determines the type of planet based on its characteristics, and then I'm gonna put these guys to the side and merge this branch, and make HOS great again 🤣

**Comment by HugoGranstrom** [comment]
Haha isn't that the most readable way to write? 🤣🤣🤣

Awesome! 😄🚀🚀🚀 Hope you don't get too gassy by all the gas giants 🤣

**Comment by TheHappyKoala** [comment]
<img width="420" alt="Screen Shot 2020-02-26 at 22 52 21" src="https://user-images.githubusercontent.com/23419647/75391640-c2157880-58ea-11ea-856e-e1b56d044583.png">

LOL. Planets with hats! Polar ice caps still need some work. Damn, I'm deep down the procedural terrain rabit hole.

**Comment by HugoGranstrom** [comment]
Wow, that's insane! 😮

**Comment by HugoGranstrom** [comment]
It looks like the Pope planet kinda😂

**Comment by HugoGranstrom** [comment]
The surface texture is so detailed and diverse!

**Comment by TheHappyKoala** [comment]
Cheers :). Yea, I'm starting to get quite happy with the results. Took a bit longer than expected, but with work I'm not able to sit down for eight hours and just blast as was the case last summer 🤣 .  Almost Friday!!!

**Comment by HugoGranstrom** [comment]
You really should be proud of yourself for what you've created 😉😊 

Haha those were the times 😂

**Comment by TheHappyKoala** [comment]
We both should be, but at the end of the day I suspect we're just doing it for the fun of it and the kick that comes with creating something... Can't think of a better motivation for doing anything in life, really. Oh, there goes wise Darrell again 🤣 

But yes, I'm naming the above planet Planet Francis. Chances are though that he will never be rendered again since he was generated procedurally.

**Comment by TheHappyKoala** [comment]
But hey, happy HOS 1 year anniversary... Think it's been a year now

**Comment by TheHappyKoala** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

**Comment by HugoGranstrom** [comment]
Wise Darrell never misses a point 😉😊 that's the truth there.

Haha it will be remembered for its short and glorious life 😂

What?!? Has it only been one year? 😮😅 It feels like we've been doing gitigans for several years together, that's crazy. Alas Happy anniversary!!! 🥳🎉🎉🎉🥂🥂

**Comment by TheHappyKoala** [comment]
I know, feel like we've been up to our gitigans for way longer than that, but it's cool to think about how far we've come in this short period of time!!! 

On the subject of progress: 

<img width="727" alt="Screen Shot 2020-02-29 at 20 16 36" src="https://user-images.githubusercontent.com/23419647/75613663-a4d1eb80-5b30-11ea-8022-0bc79fd29fd2.png">

Started with the gas giants :D Still some work to be done, but hey, I'm feeling gassy 💩

**Comment by HugoGranstrom** [comment]
That's very true, we've come a looong way since 😄

That has the potential to be really good 😉
This bad jokes fits here (or not really): it's not the fart that kills you, it's the smäll 🤦💩

**Comment by TheHappyKoala** [comment]
LOL 🤣 That made my evening, thank you!

**Comment by TheHappyKoala** [comment]
Mello is just... I mean, it's usually bad, but this is... Uhhhh 😭

**Comment by TheHappyKoala** [comment]
<img width="684" alt="Screen Shot 2020-02-29 at 21 27 25" src="https://user-images.githubusercontent.com/23419647/75614563-5d505d00-5b3a-11ea-854a-4de1d9b761ff.png">

**Comment by HugoGranstrom** [comment]
Recycling jokes must be great for the environment 😅😂

Yes it's quite terrible 😬

But that planet looks dope 😎😉

**Comment by TheHappyKoala** [comment]
Whopad00, finally committed some procedural planet stuff. Still got loads of stuff to do but go ahead and run ```npm run exoplanets``` from your terminal and watch the root get populated with textures. Still need to figure out how to do gas giants well (I want storms and stuff), but the terrestrial planets, while also in need of work, look kind of coolish.

**Comment by TheHappyKoala** [comment]
Oh, and happy spring!

**Comment by HugoGranstrom** [comment]
Happy spring to you too, Darrell! 😊☀️

Nice! Will check it out next week, I'm really busy studying for the exams 😬😅

**Comment by TheHappyKoala** [comment]
Good luck with the exams (although as always luck has nothing to do with it 😉 )!!! But will they go ahead with them? 

We're working from home now and I have to say I'm way more productive and focused at home than at the office.

**Comment by HugoGranstrom** [comment]
Thanks! 😄 Yeah I think so, we had one yesterday and unless it really becomes an outbreak at the University I doubt they will cancel them. 

Great to hear 😉 maybe you are getting a piece of your HOS productivity when you're working from home

**Comment by TheHappyKoala** [comment]
<img width="788" alt="Screen Shot 2020-03-28 at 11 25 47" src="https://user-images.githubusercontent.com/23419647/77821012-fe551800-70e6-11ea-8290-f1ea3bcfc62a.png">

I wonder what viruses might lurk on that planet.

**Comment by TheHappyKoala** [comment]
https://www.youtube.com/watch?v=wVvql19j5Cs

That cracked me up.

**Comment by HugoGranstrom** [comment]
Damn, that's sick man (🦠) 😉

Hahahaha 😂😂😂 I laugh so hard!!  Are you doing well?

**Comment by TheHappyKoala** [comment]
I'm doing alright. Considering the circumstances I genuinely have absolutely nothing to moan and groan about. Working from home is still going really well and I hope that once this is over we will have more flexibility to work remotely, as this whole experience has come to show that we're just as productive as when we go to the office in person. I'd love to be able to travel and work... I guess the only issue would be if you're in a different time zone, haha. 

First I need to pass my probation period, though. Got one month left and I've been told I do a very good job and that I fit into the team, but given the circumstances I'm not taking anything for granted. Not that concerned, though... If they decide not to keep me I'll still have a kick ass reference and companies will still need developers once the world has started turning again. 

I'm more concerned about friends that have been varslade and work in fields that are difficult to find jobs in with or without a pandemic and a failing economy. It's tough times for them. 

Bla, bla, bla 🤣 How's life treating you? Has school been suspended altogether or do you get to study remotely? How did the exams go? And what's keeping you busy if not school???

**Comment by HugoGranstrom** [comment]
Nice to hear :-) Yeah you are in luck to have a job that can be done from home. All our courses are taught at distance now so I'm home watching live stream all-day 😅 I normally commute for 2h per day but now you would expect I would have 2h more spare time but it feels like I have less than before 🤣 🤣 We had a computer laboration the other day where we were supposed to simulate a differential equation describing epidemic spread, very fitting 🤣 

Hope it works out for you and all your friends in the end 🙏😄 Stay safe!

**Comment by TheHappyKoala** [comment]
You too, Hugo... Live long and prosper! 

<img width="828" alt="Screen Shot 2020-03-28 at 20 16 23" src="https://user-images.githubusercontent.com/23419647/77831771-5add2500-7131-11ea-845f-d4de4f94ad73.png">

They are starting to look quite sexy, the gas giants, too. 

https://en.wikipedia.org/wiki/Sudarsky%27s_gas_giant_classification

**Comment by HugoGranstrom** [comment]
Wow certainly 🤩 they are becoming more and more awesome!!

**Comment by HugoGranstrom** [comment]
🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

**Comment by TheHappyKoala** [comment]
<img width="713" alt="Screen Shot 2020-03-30 at 18 49 35" src="https://user-images.githubusercontent.com/23419647/77939333-4d51a780-72b7-11ea-9b21-4d97ce3a0128.png">

**Comment by TheHappyKoala** [comment]
Hehe, sin waves are awesome!

**Comment by HugoGranstrom** [comment]
Yeah they look good! 🚀😄 Have you tried experimenting with other functions?

**Comment by TheHappyKoala** [comment]
Nothing in particular, no... Any tips? Anything that modulates is good 😄 

Cosines would be the obvious one, I guess, but I'm sure you have more options up your mathematical sleeve, so to say 🚀

**Comment by HugoGranstrom** [comment]
Hehe everything that's periodic is basically a combination of sines and cosines so my sleeves are quite empty 😅 but tan(x) would be interesting to see as it is periodic and goes from -∞ to ∞ 😂 maybe we can crash the universe with it

**Comment by TheHappyKoala** [comment]
LOL, alright, I'm gonna give that a go and report the results provided there's still a universe to report on 🤣

**Comment by HugoGranstrom** [comment]
Haha 😂😂😂😂

**Comment by TheHappyKoala** [comment]
How ya doing, bra? When does school wrap up?  And how's  studying remotely working out for you?

I'm good. Feel very lucky because I passed my probation period and I am now tillsvidareanställd 😄 . And it's kind of bizare because we are korttidspermiterade, so I only work 60% of the hours, but I get 94% of my normal salary.

**Comment by HugoGranstrom** [comment]
I'm doing very well, feeling more "pig and cry" then ever before 😂 I do also get to sleep like 2h more now when doesn't have to commute to school any more, that's really nice 🤩😂😂😂 Our schedule hasn't really changed so in the beginning of June my last exams takes place. But I should be honest and say that I'm quite bored nowadays just being home and not meeting my classmates. But I try to fill the time with stuff to do. 

The fluid dynamics exam went well so in theory I should have a better grasp of the SPH stuff now but I'm lacking the motivation 😬 it just seems so big and most of all, hard to test if it works 🤔 

Woho! 🥳🚀 Congrats! They must really like your work then 😄 how do you pass your time when your not working? Going out buying "smoked eel"? 😜

Yeah this entire situation is quite bizarre and I'm wondering how we will look back at it in the future.

**Comment by TheHappyKoala** [comment]
Hahaha... Smoked eel? Is that a thing now?? 

Done some work on HOS, but most of the time has been spent on reading and working out. Being home pretty much all the time has made me really appreciate exercise. 

But yea, like you miss hanging out with your class mates, I miss going to the office and seeing my colleagues. We do a virtual fika, but it's not the same. 

First day back at the office, whenever that will be, will be a very happy day. I miss simple every day stuff like the capucino I buy at Pressbyrån on my way to the office, and going out for lunch. I know you can still do that, but it feels like an unnecessary risk to take these days. 

Life is beautiful, though, and I have much to be grateful for, so yay 😄 

Apparently Newton wrote Principia while in qurantine... Tell me, are you composing the next chapter in our understanding of the Universe while you're homebound 🤣 ? It's so easy to feel like a docile bum when you read about all these great minds and what they had done before the age of 23. 

Happy Sunday 😄

**Comment by HugoGranstrom** [comment]
There was a new food hall that opened recently and there were a lot of old people that went there on the opening and an old lady has become sort of a meme for saying she just wanted to buy smoked eel 😂😂😂

That sounds like good things to do, I should probably workout a bit more myself. I try to go for a walk every day but that's it. 

There are so many things we take for granted and don't appreciate as much as we should when we have them, like your capucino. I hope we do appreciate them more when we get back to normal. 

Hahaha 😂😂😂 Newton is on an entirely different level. Imagine if he had had the internet back then, would he still have done all the amazing discoveries then or would he instead have spent time watching cat movies? 🤔😂😂😂 Yeah you feel a bit lame when you compare yourself to the geniuses 😅 but on the other hand, now you have the ability to learn more stuff then they ever knew thanks to the accumulation of knowledge in the 300 years since he was running about in the world 😄

**Comment by TheHappyKoala** [comment]
How about we make RKN64 the new default integrator? Having pefrl as the default feels like having a ferrari but choosing to drive a volvo v70 instead...

**Comment by TheHappyKoala** [comment]
Damn it, let's go for RKN12

**Comment by HugoGranstrom** [comment]
Haha 😂😂😂😂😂😂 yeah why take the Ferrari when you have a turbojet.

**Comment by HugoGranstrom** [comment]
Why did we change it to PEFRL back then?

**Comment by TheHappyKoala** [comment]
I was obsessing over mobile devices, but rkn12 actually runs rather well on those, too

**Comment by HugoGranstrom** [comment]
The best experience is doubtlessly the RKN12 so let's go with it! 🚀🚀🚀🚀🚀

**Comment by TheHappyKoala** [comment]
Yaaayyyy 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

**Comment by HugoGranstrom** [comment]
Unless we are using the largest allowed timestep generally higher order methods are more efficient and should be preferred on weaker devices like phones

**Comment by TheHappyKoala** [comment]
Yea, you're absolutely correct in noting that... And usually it's more fun when things happen faster rather than having to wait for Uranus to stroll like an old man with a rullator through the solar system

**Comment by HugoGranstrom** [comment]
Haha yeah definitely!😂😂🚀🚀🚀🚀 Unless you want to give a slow-TV experience of course 😜

**Comment by TheHappyKoala** [comment]
Should mention both of those textures were generated procedurally!

**Comment by HugoGranstrom** [comment]
Happy June to you too! 😊 I'm doing fine, I had my last exam for the semester today so nearly summer break. I just have one assignment left before I can take a break. Later this summer I will take two courses in Astronomy which will be fun and hopefully I will learn something that could be of use here as well 😄 

How are you doing?

Wow wow wow!!!🤩 Those are awesome! 🚀🚀🚀🚀 I couldn't tell that wasn't an artist who made it with the amount of detail it has. You have done a splendid job Darrell! 😁🚀 Wow!

**Comment by TheHappyKoala** [comment]
Oh, that's awesome. Do you have any idea what the courses will cover??? 

I'm doing alright. Happy that the weather is a bit more agreeable now, and work is really interesting, too. And I'm having bucket loads of fun working on the procedural planet textures in my spare time. Made some revisions to the GUI, as well, so I'm quite stoked about getting all of these new features out there. 

It's funny, but with Google Analytics I can see that there's somebody from Houston that uses our simulator pretty much every day, and he seems to be very interested in exoplanets, and I  get all giddy and happy knowing that the new planet textures will probably make this person very happy. It's an awesome feeling when you know that you give people something that they find interesting.

**Comment by HugoGranstrom** [comment]
I think it's more focused on the qualitatively than quantitatively parts of astronomy. One of them sounds like it's the history of the universe basically and the other is more focused on what's out there now. 

That's great to hear! I'm looking forward to seeing the results in my own computer. I've recently bought a new laptop so now it should turn into a meltdown every time I run HoS 😂

That's some dedication! 🤩 I'm sure that person will be positively surprised when one day it has changed completely to a new cool look! 😄

**Comment by TheHappyKoala** [comment]
Woho! What are the specs of your new bad boy of a computer? 

By the way, check out this scenario: https://gravitysimulator.org/spaceflight/operational-spacecraft-in-orbit-around-mars

It's a simulation of all spacecraft in orbit around Mars as of today!

**Comment by HugoGranstrom** [comment]
They aren't too impressive but a lot less unimpressive than my old at least. The new Ryzen 4500U with 6 cores and 8GB RAM vs my old core i3 4030U with 4GB RAM. It's performing really well thus far. I have played a few games and the fans are barely noticeable now. 

Oh la la! 🤩 I like! They are so close to the surface! 🌐 I tried switching to RKN12 but it seemed like it wanted to use an even lower timestep, so I guess that's because of the small orbits there. I didn't realize there were so many spacecrafts around Mars. 😮😀 And Phobos and Deimos are so close to it!

**Comment by TheHappyKoala** [comment]
Phobos will actually collide with Mars in a couple of million years... Its orbit is deteriorating since its orbital period is shorter than the rotation period of Mars 😱 . Deimos, on the other hand, will eventually escape Mars's gravity. 

Building textures for 4000 + exoplanets takes quite a while, so I'm gonna go about it in a piece meal fashion, which is why I'm asking you, Hugo, esteemed collaborator and developer extraordinaire - do you have any favourite exoplanetary systems that you would like textures for 😄 ? If you don't, google exoplanets and pick a couple of systems 🤣

**Comment by HugoGranstrom** [comment]
That's kinds crazy to think about 😵 our own Moon is also escaping us if I don't remember wrong. It's so weird thinking about how the future will look so different! 😱 

Are you building the textures manually? 😬 Sounds like alooot of time. 

Haha I like that title 😂 Oh Darrell, the master of the harmonic spheres and master of the space realm, I will come back later tonight when I have watched Midsomer Murders and have looked up a random Wikipedia page on Exoplanets 😉

**Comment by TheHappyKoala** [comment]
Haha, no!!! Written some code that does it for us based on mass, radius and temperature data for the exoplanets! Making a commit in a bit. So you run npm run exoplanets and then it spits out the textures into the textures directory in the static directory. It's all procedural. 

On the subject of British crime dramas... I can warmly recommend Kommisarie Lewis 😄 

Well, our Moon won't escape... It will recede until Earth and the Moon are tidally locked... Earth is giving its rotational energy to the Moon which makes it recede away from Earth

**Comment by TheHappyKoala** [comment]
Sounds good, though!

**Comment by TheHappyKoala** [comment]
Building one 3000 x 3000 px texture takes like 5 mins, which is why I'm not gonna do them all at once, by the way 🤣

**Comment by HugoGranstrom** [comment]
Ahh that sounds more reasonable 🤣👌 thank god! 

Oh I will check it out for sure! You're a bit of a British crime drama expert if I recall correctly ;)

**Comment by HugoGranstrom** [comment]
Seems like I have misunderstood our Moon then 😅🤣 good thing I'm taking the astronomy course now!

**Comment by HugoGranstrom** [comment]
I've found https://en.m.wikipedia.org/wiki/List_of_exoplanet_extremes and the subsection System Characteristics seems interesting: Kepler-90, Trappist-1, HR-8799, Kepler-444, Kepler-52,  HD 20782.

Räcker det?

**Comment by HugoGranstrom** [comment]
Ooops was writing with a friend in Swedish in parallel 😂

**Comment by TheHappyKoala** [comment]
Det räcker! Fy fasiken vad varmt det var idag! 🌞 

That's a fine choice of exoplanetary systems right there, and awesome list! 

I'll get right to it 🚀

**Comment by HugoGranstrom** [comment]
Haha 🤣 yes it's too damned warm! I'm up in the mountains fishing currently and it's too hot even up here. It's a choice of either being stung by mosquitoes or sweating like a pig when you choose the clothes 🤣😂😂

Awesome! I'll look forward especially for those! 🤩

**Comment by HugoGranstrom** [comment]
🚀

**Comment by TheHappyKoala** [comment]
Turn off the habitable zone and labels, and then you can inspect the planets of the Trappist 1 system by changing the camera focus: https://gravitysimulator.org/exoplanets/trappist-1

I use a conservative estimate for the habitable zone and according to that calculation only e is in the habitable zone. 

Still loads of work to do to make exolplanet scenarios more interactive and informative, but this is most certainly a start 😄

**Comment by HugoGranstrom** [comment]
That's really cool to see the transition between close, habitable zone and far away planets! 🚀🚀🚀 Great work!! Your well on your way to a freaking awesome exoplanet explorer!

**Comment by TheHappyKoala** [comment]
Okay, turns out my habitability zone calculations only took stars of the Sun's mass into account; now they take the whole spectrum of stellar masses into account 🤣 

Spent yesterday cross checking my simulations with pictures of various exoplanetary habitable zones and now things look pretty good now in that regard. Anyhow, new Trappist 1 textures: https://gravitysimulator.org/exoplanets/trappist-1. Also added textures for the other systems you suggested in addition to some 20 other systems 🚀 

And note the cool camera tweening effect when you click on display habitable zone or display barycenter... Smooth.

**Comment by HugoGranstrom** [comment]
Ohhhh that animation is slick! 😁🚀🚀🚀

Awesome! I will play around with it tomorrow before my astronomy lecture!

**Comment by HugoGranstrom** [comment]
Been playing around a bit now and I noticed something curious when I played around with the Physics settings in the Trappist-1 scenario. If you turn on Barnes-Hut the planets just continue in a straight line. I couldn't replicate this in any other scenario strange enough 🤔

**Comment by TheHappyKoala** [comment]
Ahhh... Scenario had theta set to 0.5, which is much too high. Lowered it and then it works as expected. Is there a formula for calculating an appropriate value for theta given the spatial separation between bodies?

**Comment by TheHappyKoala** [comment]
Trappist-1 is a super tight system, so that's why the 0.5 value for theta became an issue... Find another tightly packed red dwarf system and you're bound to see the same issue, because I think that for exoplanetary systems, theta is always set to 0.5 by default

**Comment by HugoGranstrom** [comment]
Hmm, I don't know about that, because even if I set it theta = 0.1 they just continue in straight lines, only theta = 0 seems to be stable. Did you manually set theta to something between 0 and 0.1 and got it working? Even if theta is set too high it should only make the acceleration calculations less accurate, not make all accelerations zero like it seems here :/ But it's odd that it only seems to appear in small systems. 

I think I'm ready to get back into coding a bit more on HOS now that my fishing trip is done and I've caught up to my school stuff finally. But I barely remember what I left off 🤣🤣

**Comment by HugoGranstrom** [comment]
I found me last message related to this issue before I took me break 🤣  https://github.com/TheHappyKoala/Harmony-of-the-Spheres/issues/115#issuecomment-583855285

**Comment by HugoGranstrom** [comment]
Think you've got enough time to think about it now? 😜

**Comment by TheHappyKoala** [comment]
Lol, I'd totaly forgotten what this thread was originally about

**Comment by TheHappyKoala** [comment]
But sure, it's ready for a pull request?

**Comment by TheHappyKoala** [comment]
Would be super sweet, because then we can move forward with the lambert stuff in earnest

**Comment by TheHappyKoala** [comment]
I set theta to 0.05

**Comment by TheHappyKoala** [comment]
And did you catch any big fish?

**Comment by HugoGranstrom** [comment]
Size doesn't matter, right? 🤣 I caught about 30 trouts and some of them could be considered medium sized perhaps....

I don't think I did anything after posting that message back in february so the answer is no..... 😅

**Comment by TheHappyKoala** [comment]
Damn... Don't think I've ever laid my eyes on that many trouts in my whole life! Well, is this something you'd like to work on??

**Comment by TheHappyKoala** [comment]
I'm about to make the filthiest commit of all time... I just want this branch to be history

**Comment by HugoGranstrom** [comment]
Hmm the value of theta shouldn't have to be that small though, it should as a worst-case scenario take the center of mass as a body but it seems like it is complety ignoring that 🤣

**Comment by TheHappyKoala** [comment]
Done... Look at the latest commit to the gatsby branch 🤣 

Cool thing, though, is that the moment I make the commit, a build is run for the production website 🚀

**Comment by TheHappyKoala** [comment]
Oh shit, uploaded a layout bug.... Another commit coming right up

**Comment by HugoGranstrom** [comment]
Haha you could perhaps try to fish from a bridge in town ;) maybe you could get a *yummy* pike.... 

I'll get straight back to doing what I have said I'd do 😄 The two functions should be done this week if I can get up to speed with how the code works again. 

`Showing 1,272 changed files with 22,333 additions and 9,054 deletions.` Holyshmoly! That's indeed the filthiest commit I have ever seen in my git-days! 🤣

**Comment by HugoGranstrom** [comment]
Seems like you got it fixed! 😃 Does it specifically build from the `migrate_to_gatsby` branch or any branch in the repo? In other words, do I need to be afraid when I merge stuff into master 🤣 😱

**Comment by TheHappyKoala** [comment]
Haha, well, it's a horrible commit, but not quite that horrible, as all exoplanet scenarios are generated with the npm run exoplanets script 🤣 

Working on breaking up the exoplanet scenarios into more manageable categories, though: https://gravitysimulator.org/exoplanets/anglo-australian-telescope

So far each observatory is one category, but I'm gonna add a category for multiplanetary systems and perhaps even more interesting one for systems with planets that could potentially be habitable 🚀

**Comment by TheHappyKoala** [comment]
It builds from migrate to gatsby at the moment

**Comment by TheHappyKoala** [comment]
<img width="1352" alt="Screenshot 2020-07-01 at 19 49 05" src="https://user-images.githubusercontent.com/23419647/86275708-f8823d80-bbd3-11ea-89dc-c9c641cb6a04.png">

**Comment by TheHappyKoala** [comment]
It's like the awesomest thing ever... Push and it builds

**Comment by HugoGranstrom** [comment]
Haha don't know if I dare run that script tbh 🤣 maybe my new computer explodes or something.

Yeah the exoplanets need every way to make it more manageable 😄 The thing that would make the most difference is probably "just" a search function so people can find HD 102195 without knowing to go to the Kitt Peak National Observatory. Don't you think? 

Adding more interesting "explore" categories sounds like a really nice idea as well! 🚀

**Comment by TheHappyKoala** [comment]
Yea, that's a good idea... A search page where you can query scenarios with a string and maybe add filters to narrow things down

**Comment by HugoGranstrom** [comment]
Wow! That's so cool, built-in to Github and all 😮

**Comment by TheHappyKoala** [comment]
Yea, so all you need to do is make a commit and it will do the rest

**Comment by TheHappyKoala** [comment]
So no worries, and reverting to old builds is super easy

**Comment by HugoGranstrom** [comment]
That's really convenient I must say but also a bit frightening if one happens to commit something that shouldn't be commited. But I guess we could have a separate production and development branch in the future?

**Comment by TheHappyKoala** [comment]
Yea, in the future the idea is to have the builds run from master, and then we have feature branches that we merge into master... This is just a temporary setup as what's on the master branch is basically 2 billion years old

**Comment by HugoGranstrom** [comment]
Haha yeah, it's been left there for a while now... But I noticed we've reached 1000 commits now! 🥳

**Comment by TheHappyKoala** [comment]
Oh... True 🥳 !!! Thanks for hanging around for this long 😄

**Comment by HugoGranstrom** [comment]
It's been a pleasure 😄 and hanging is basically all I've done recently XD

**Comment by HugoGranstrom** [comment]
It's amazing how much time and effort you have put into HOS during the years 😮 at least 900 of those commits must have been you after all. 

Have you seen if the guy from Houston is still hanging in HOS a lot still?

**Comment by TheHappyKoala** [comment]
Yups, he was online today. There's a guy from Florida that keeps on coming back, too. Our biggest fans 😄 !

True, but your contributions have been priceless. Without your contributions there would be no high-fidelity integrators, lamert solvers or collisions since they are not possible without an integrator that allows for a variable time step

**Comment by TheHappyKoala** [comment]
But still, in June HOS turned two, so it hasn't been that long... We've done quite a bit in two years time 😄 .

**Comment by TheHappyKoala** [comment]
Feel free to revert to one of the first iterations from June 2018 🤣

**Comment by TheHappyKoala** [comment]
I was still living in the Netherlands at the time

**Comment by HugoGranstrom** [comment]
Woho! 🥳 Soon we'll have a fan club of some sort. Except they don't know they are in it. 🤣 

Thanks man! 😄 Teamwork is the key to success! Oh when you put it that way, we've done a looooot of stuff, just comparing the master branch with the gatsby branch is quite a difference UI-wise. Haha I must try the old version someday 🤣 maybe we could make a nostalgia-version as an easter egg of some sort? 😂

**Comment by TheHappyKoala** [comment]
Sure is 🚀 !

But yea... That would be fun, Euler integrator and the whole shebang 🤣 

But wiiieee... We now have textures for all systems with two or more planets: https://github.com/TheHappyKoala/Harmony-of-the-Spheres/commit/04d98fa9cd0d2e22bc7f375b27efb436da207f0f

Made them a bit more light weight so that they load faster, as well. Running a job now for the remaining 2000 or so textures 😅 !

Started watching Midsomer Murders, by the way... It's a delightfully hilarious show 🤣 ! I always thought it was just a crime who done it show, but every episode cracks me up, pretty much.

**Comment by HugoGranstrom** [comment]
Woho! The textures are taking over the web! I've experienced a bit of a slow start when loading scenarios so that may be a good idea :-) 

Yeah I now, it nearly feels more like a humor program than a criminal drama sometimes. 🤣 

I've managed to make a somewhat working Orbital elements integrator now that works for elliptic orbits (given there is a body named "Sun" in the simulation, haven't really looked at how hard it would be to fix yet). Two questions:
- Do you think we should try to make it work for parabolic and hyperbolic trajectories as well? (Borisov for example doesn't work with my current implementation) It should be possible to add this in the future without any changes to any existing code except to the `stateToKepler` and `keplerToState` functions internally. 
- As you already use RKN12 for trajectory crunching, would it be possible to just switch to this integrator? Or do you think a specialized function is needed to get the future position of just one specific body instead of doing it for all? I don't know how much faster it would be but we'd have to traverse the SOITree in search for that specific body in this case and I haven't really come up with a non-hacky way of doing it yet. (not a hacky way either for that matter 🤣 )

I'd love to hear your opinion on this 😄 In the "best" scenario I'm basically ready to create a pull request (after I have fixed any merge conflicts, it's based on the master branch...).

**Comment by TheHappyKoala** [comment]
Yea, in the previous iteration, the textures were 3.5 mb each, and now they'r between 200 and 500 k, and to be honest quality wise there's not much of a change, or maybe I need to get myself a pair of glasses 🤣 

Very good questions. Here's my take:

 1. Yes, it would be excellent if parabolic and hyperbolic trajectories were supported. First because, as you correctly point out, objects like Borisov and Oumuamua do not have closed orbits with respect to the Sun (I suppose they do with regard to the Milkyway, though?). Second, because when simulating space craft trajectories, quite often you will find yourself in a situation where, with respect to the primary, for example Earth, the spacecraft is on a hyperbolic trajectory, i.e it's leaving the system. But also for the sake of simulating actual spaceflight missions like New Horizons, Voyager and so on, we would need that capability. So if it's not too much bother for you, that would be super duper awesome 😄 

2. Not sure I follow you... Could you provide a use case illustrating what you have in mind? 

Oh... I really need to get the branches sorted, it's unfair that I'm making you resolve conflicts because of my gitigans 🤣

**Comment by HugoGranstrom** [comment]
That's quite a big reduction :O 

1. Then I'll make that my next mission, fixing the hyperbolic and parabolic orbits work. 😄 But I guess we would never simulate anything we want an accurate result from with orbital elements? I mean, we will do a quick prediction of where our destination planet will be in the future with orbital elements, but for the actual simulation of the spacecraft's motion we will use RKN12, right? But it's nice if we have the option to use orbs as well of course.
2. Right now the trajectorycruncher uses RKN12, could we just replace it with the orbital elements integrator now (but we just have to take on timestep instead of 1000s)? Ignore the rest of my question 🤣

**Comment by TheHappyKoala** [comment]
Gonna get back to you in a few... I had that rare thing in the life of Darrell, a weekend with family and friends 🤣

**Comment by HugoGranstrom** [comment]
That sounds awesome! ☺️ Hope you enjoyed it to it's fullest
